src/string_id.h

```diff
@@ -69,69 +71,71 @@ template<typename T>
 class string_id
 {
     public:
+        template<class U>
+        friend class string_id;
         using This = string_id<T>;
 
-        /**
-         * Forwarding constructor, forwards any parameter to the std::string
-         * constructor to create the id string. This allows plain C-strings,
-         * and std::strings to be used.
-         */
-        // Beautiful C++11: enable_if makes sure that S is always something that can be used to constructor
-        // a std::string, otherwise a "no matching function to call..." error is generated.
+        // Constructor for compile time strings (Compile time known length)
+        template<unsigned N>
+        explicit constexpr string_id( const char ( &str )[N] ) {
+            _string = insert_or_get_cache( str, false );
+        }
```

Comment:
[Repeating this comment anew because the old version got marked outdated]

This is quite dangerous; you're assuming that any `char[]` used to construct a `string_id` is a string literal (or at least is a thing with static lifetime).  But short-lived `char[]` objects can exist, and then your `string_id` would end up with a dangling pointer.

See for example [this code](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:3,positionColumn:2,positionLineNumber:3,selectionStartColumn:2,selectionStartLineNumber:3,startColumn:2,startLineNumber:3),source:'template%3Cunsigned+N%3E%0Aconstexpr+void+f(+const+char+(+%26str+)%5BN%5D+)%0A%7B%0A%7D%0A%0Avoid+g()%0A%7B%0A++++char+c%5B10%5D%3B%0A++++f(c)%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:g102,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:c%2B%2B,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'x86-64+gcc+10.2+(Editor+%231,+Compiler+%231)+C%2B%2B',t:'0')),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,wrap:'1'),l:'5',n:'0',o:'%231+with+x86-64+gcc+10.2',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

Unfortunately, there's no way to detect lifetimes in C++ so you can't make this really safe.  We *could* add a clang-tidy check to verify that this constructor was only ever called on actual string literals, so it's not an insurmountable issue.